src/server/ua_services_session.c

```diff
@@ -186,56 +166,27 @@ Service_CreateSession(UA_Server *server, UA_SecureChannel *channel,
-       channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT) {
-        const UA_SecurityPolicy *const securityPolicy = channel->securityPolicy;
-        const UA_ByteString *const localCertificate = &securityPolicy->localCertificate;
-
-        UA_ByteString dataToVerify;
-        UA_StatusCode retval = UA_ByteString_allocBuffer(&dataToVerify,
-                                                         localCertificate->length + session->serverNonce.length);
-
-        if(retval != UA_STATUSCODE_GOOD) {
-            response->responseHeader.serviceResult = retval;
-            UA_LOG_DEBUG_SESSION(server->config.logger, session,
-                                 "Failed to allocate buffer for signature verification! %#10x", retval);
-            return;
-        }
-
-        memcpy(dataToVerify.data, localCertificate->data, localCertificate->length);
-        memcpy(dataToVerify.data + localCertificate->length,
-               session->serverNonce.data, session->serverNonce.length);
-
-        retval = securityPolicy->certificateSigningAlgorithm.
-            verify(securityPolicy, channel->channelContext, &dataToVerify,
-                   &request->clientSignature.signature);
-        if(retval != UA_STATUSCODE_GOOD) {
-            response->responseHeader.serviceResult = retval;
-            UA_LOG_DEBUG_SESSION(server->config.logger, session,
-                                 "Failed to verify the client signature! %#10x", retval);
-            UA_ByteString_deleteMembers(&dataToVerify);
-            return;
-        }
-
-        retval = UA_SecureChannel_generateNonce(channel, 32, &response->serverNonce);
-        UA_ByteString_deleteMembers(&session->serverNonce);
-        retval |= UA_ByteString_copy(&response->serverNonce, &session->serverNonce);
-        if(retval != UA_STATUSCODE_GOOD) {
-            response->responseHeader.serviceResult = retval;
-            UA_LOG_DEBUG_SESSION(server->config.logger, session,
-                                 "Failed to generate a new nonce! %#10x", retval);
-            UA_ByteString_deleteMembers(&dataToVerify);
-            return;
-        }
-
-        UA_ByteString_deleteMembers(&dataToVerify);
-    }
+static UA_StatusCode
+checkSignature(const UA_Server *server, const UA_SecureChannel *channel,
+               UA_Session *session, const UA_ActivateSessionRequest *request) {
+    if(channel->securityMode != UA_MESSAGESECURITYMODE_SIGN &&
+       channel->securityMode != UA_MESSAGESECURITYMODE_SIGNANDENCRYPT)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    if(!channel->securityPolicy)
+        return UA_STATUSCODE_BADINTERNALERROR;
+    const UA_SecurityPolicy *securityPolicy = channel->securityPolicy;
+    const UA_ByteString *localCertificate = &securityPolicy->localCertificate;
+
+    size_t dataToVerifySize = localCertificate->length + session->serverNonce.length;
+    UA_ByteString dataToVerify = {dataToVerifySize, (UA_Byte*)UA_alloca(dataToVerifySize)};
```

Comment:
Baaaad idea. Certificates can be 500-1000 bytes size. Allocating that on the stack isn't a good idea imho. Furthermore it's user supplied data and the size can be set to any desired length.

Reply:
Okay i just saw the check you added. I don't think its a good idea to limit the size to 4096.
Why are we not using malloc here? I see no reason not to.

Reply:
- I hate branches (if/else)
- Removing branches makes me happy
- Alloca requires less branches

I added a TODO that refers to MaxSenderCertificateSize from part 6.
If you want to add that or a solution with malloc, feel free to do so.